various: correct arguments to k_sleep

pabigot opened the pull request:
The `k_sleep()` kernel function is documented to take a value in milliseconds.  A significant number of in-tree calls to it first pass the value through `K_MSEC()`, which is intended to convert milliseconds to a timeout for kernel API such as `k_poll()` and `k_sem_take()`.  Currently this macro does nothing, while its companions like `K_SECONDS()` convert their values to milliseconds.

In the future `K_MSEC()` and its kin will produce an opaque timeout value that supports precision finer than 1 ms and alternative clocks. In that future `k_sleep()` will continue to take its argument as milliseconds (as `k_usleep()` will continue to use microseconds), and passing a timeout value to `k_sleep()` will result in compilation errors.

Replace the arguments to `k_sleep()` that involve `K_MSEC`-like macros with values that provide milliseconds.

(Several of these are necessary to get #17155 to pass sanity-check; others are unexercised by the test infrastructure.)

pabigot committed:
passim: correct arguments to k_sleep

The k_sleep() kernel function is documented to take a value in
milliseconds.  A significant number of calls to it first pass the
value through K_MSEC(), which is intended to convert milliseconds to a
timeout for kernel API such as k_poll() and k_sem_take().  Currently
this macro does nothing, while its companions like K_SECONDS() convert
their values to milliseconds.

In the future K_MSEC() and its kin will produce an opaque timeout
value that supports precision finer than 1 ms and alternative clocks.
In that future k_sleep() will continue to take its argument as
milliseconds, and passing such a value to k_sleep() will result in
compiler errors.

Replace the arguments to k_sleep() that involve K_MSEC-like macros
with values that provide milliseconds.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

jhedberg commented:
Is the reasoning documented somewhere why `k_sleep()` isn't converted to use `k_timeout_t`? I'd have thought that to be the simplest thing to do, since it avoids having multiple different "sleep" APIs in the kernel.

jhedberg commented:
Another comment: it's not really nice how the code now becomes less readable. Previously you could see the logical duration with the help of these macros, but now someone not familiar with the kernel APIs has to go and look up the k_sleep definition/documentation to understand how the integer passed to it gets interpreted.

pabigot commented:
@jhedberg I don't think it was documented anywhere; we're only barely starting to take notes on lower-level telecons, and haven't yet come up with a way to confirm and record consensus positions.

The compelling argument for me was consistency with standard usage: `sleep(3)` takes its parameter in milliseconds and `usleep(3)` in microseconds.  So Zephyr's functions should follow that existing practice.

I expect, if the timeout concept gets taken to its logical conclusion, that there will be a `k_delay(timeout)` function added that is about the same as `k_sleep()` but for any timeout specification and clock source.

pabigot committed:
tests/threads/thread_apis: exclude platforms failing on shippable

These three platforms build an image that does not fit into flash,
causing the test run to fail--but only on shippable.  They aren't run
with local sanitycheck.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

jukkar commented on drivers/dma/dma_stm32f4x.c:
```diff
@@ -275,7 +275,7 @@ static int dma_stm32_disable_stream(struct dma_stm32_device *ddata,
 
 		/* After trying for 5 seconds, give up */
-		k_sleep(K_SECONDS(5));
+		k_sleep(5 * MSEC_PER_SEC);
```
At one point in the past we had things like this `sec * MSEC_PER_SEC`, then `K_SECONDS()` was introduced and code was converted to use it, and now we are planning to go back. I would actually prefer the `K_SECONDS()` here as it is much more intuitive to use.

pabigot replied:
I agree; the problem is it's being mis-used here.

`K_NO_WAIT`, `K_MSEC(x)`, `K_SECONDS(x)` are all currently documented to return a timeout value.  It happens that in the past all timeouts were in milliseconds, but that's no longer sufficient.  So the proposal on the table is to *some* but not all API that currently expects milliseconds (viz. `k_timer`, `k_poll`, but not `k_sleep`) to instead take a duration representation with better resolution.  This means API that produced timeouts will no longer produce a value that's suitable to `k_sleep()`.

There's no reason other than it's ugly that we can't have something like `K_MSEC_unit()` to indicate the translation is to milliseconds.

Or we could convert to C++ where all this would be much easier: `k_sleep(5_s)`

Regardless, whether `k_sleep()` switches to a timeout or remains milliseconds is a decision we'll have to consider.  I'll put this PR on hold until we get there.

andyross commented:
Sorry, I missed the discussion here.  Honestly from my perspective the use of a timeout macro to k_sleep/k_usleep was always sort of a bug.  These functions do not take consistent units.  I mean, this is legal in the current tree too, but totally wrong:

    k_usleep(K_MSEC(1)); / *sleeps for 1us, not 1ms! */

The macros just don't do what is being requested, and never did.

But now, for the case of k_timer and other things taking a k_timeout_t argument, we **do** have the ability to be strict about units.  But sleep/usleep don't work that way for historical reasons (because sleep/usleep on unix don't, mostly).

jukkar commented:
> These functions do not take consistent units. I mean, this is legal in the current tree too, but totally wrong:

This sounds then like a good motivation to change the parameters to be consistent units i.e., to a generic timeout value.

pizi-nordic commented:
> This sounds then like a good motivation to change the parameters to be consistent units i.e., to a generic timeout value.

I fully support this approach. We should have same units in whole kernel API.

pabigot commented:
> @pabigot What is the "passim" in the PR and commit title?

It's used in [academic references](https://www.lexico.com/en/definition/passim) to mean "to be found at various places throughout the text."

I use it when indicating that a common change is made throughout the entire tree, where there is in principle no bounding scope.  This change touches drivers, lib, samples, subsys, and tests and there's no real motivation to separate them, so "passim" serves as the subsystem name.

I'm unaware of an existing standard for this case.

jhedberg commented:
> sleep(3) takes its parameter in milliseconds

Where do you get that from? POSIX defines sleep as taking seconds: https://pubs.opengroup.org/onlinepubs/009695399/functions/sleep.html

I've personally never made the association of Zephyr's `k_sleep` with this, so I'd prefer us to have just a single sleep API and keep its timeout parameter type consistent with the rest of the kernel APIs.

pabigot commented:
Per API review 2019-09-24 the plurality opinion is that `k_sleep()` should take a timeout, which means that all passing of integers to this routine should switch to using macros.

Drop this until we confirm that #17155 will progress to the point where this actually matters.

pfalcon commented:
> >    @pabigot What is the "passim" in the PR and commit title?

> It's used in academic references to mean "to be found at various places throughout the text."

I thought it's some kind of "passive simulator", another thingy along the lines of native_posix/nrf52_bsim. It would be nice to stick with software development terminology, and many projects use "various: " in this context. Actually, let me lead the example and change the title of this PR.
